### DISCUSSION TITLE
How to send diff tokens with wagmi and metamask ?

### DISCUSSION THREAD

**Discussion by victorpavlenko** [wevm/wagmi#2279] [Q&A] [answered]
I configured everything exactly according to the docs, but in my case the user can send:

- **ETH**
- **USDT:BEP20**
- **USDT:ETH20**
- **USDC:BEP20**
- **USDC:ERC20**
- **USDC:MATIC**

In general, I setup all chains `[mainnet, bsc, polygon]`, I also made auto-detection of the network, depending on the selected blockhain. 
And what about the token? Should I change something else besides **decimal** to send for example USDT? 
Or just update chain and decimal? 

PLZ HELP This is my first encounter with Blockchain.

```
const { crypto_address, crypto_amount, gateway, blockchain } = invoiceInfo;

const currentCrypto = (gateway === "USDT" || gateway === "USDC") ? CRYPTO[gateway][blockchain] : CRYPTO[gateway];
    
const [debouncedTo] = useDebounce(crypto_address, 500);
const [debouncedAmount] = useDebounce(crypto_amount, 500);
    
const { chain } = useNetwork()
const { chains, pendingChainId, switchNetwork } = useSwitchNetwork()

const { config, error: prepareError } = usePrepareSendTransaction({
    request: {
        to: debouncedTo,
        value: debouncedAmount ? utils.parseUnits(`${parseFloat(debouncedAmount)}`, currentCrypto.decimals) : undefined,
    },
});

const { error, sendTransaction } = useSendTransaction(config);

const { address, connector, isConnected } = useAccount();

const { data: ensName } = useEnsName({ address });
const { connectAsync, error: connectedError, connectors, isLoading, pendingConnector } = useConnect();

const data = useConnect();
const { disconnect } = useDisconnect();

const { isLoading: LOADING, isSuccess } = useWaitForTransaction({hash: data?.hash, });

```

**Comment by denvradiy** [comment]
Hi! 

did you solve your issue?

**Reply by victorpavlenko** [reply]
Yes I did, but I was just starting to figure it out back then so I can't even understand my own question now lol. 
Everything is extremely simple.

**Reply by victorpavlenko** [reply, marked as answer]
```
address: currentCrypto.contract 
abi: currentCrypto.abi
functionName: "transfer"
args: [addressTo, amount]
```

This is config to prepare contact for any token u need.
You can find contact address and abi for example on etherscan.io

**Reply by DimeRoyal** [reply]
Hello, Can you tell me what is the issue here, I am trying to make a transaction of USDT. Plz help
Error
Transaction failed: ContractFunctionExecutionError: Missing or invalid parameters. Double check you have provided the correct parameters. URL: https://cloudflare-eth.com Request body: {"method":"eth_call","params":[{"data":"0xa9059cbb000000000000000000000000051acf8d69a18d6a6e4dd063b39501e69704ac5000000000000000000000000000000000000000000000000000000000000003e8","from":"0xef7586E34A69E261Aa313EBF360DC0510455fd05","to":"0xdAC17F958D2ee523a2206206994597C13D831ec7"},"latest"]} Raw Call Arguments: from: 0xef7586E34A69E261Aa313EBF360DC0510455fd05 to: 0xdAC17F958D2ee523a2206206994597C13D831ec7 data: 0xa9059cbb000000000000000000000000051acf8d69a18d6a6e4dd063b39501e69704ac5000000000000000000000000000000000000000000000000000000000000003e8 Contract Call: address: 0xdAC17F958D2ee523a2206206994597C13D831ec7 function: transfer(address to, uint256 amount) args: (0x051ACf8d69a18d6a6e4Dd063B39501e69704Ac50, 1000) sender: 0xef7586E34A69E261Aa313EBF360DC0510455fd05 Docs: https://viem.sh/docs/contract/simulateContract Details: invalid opcode: INVALID Version: viem@2.21.44

`const USDT_ABI = [
  {
    type: 'function',
    name: 'transfer',
    stateMutability: 'nonpayable',
    inputs: [
      { name: 'to', type: 'address' },
      { name: 'amount', type: 'uint256' },
    ],
    outputs: [{ name: '', type: 'bool' }],
  },
];

const usdtContractAddress = '0xdAC17F958D2ee523a2206206994597C13D831ec7';

const recipient = ref('');
const amount = ref(0);

const { connect } = useConnect();
const { address, isConnected } = useAccount();
const connected = computed(() => isConnected.value);

const {
  writeContractAsync,
  isPending,
  isSuccess,
  isError,
  failureReason,
} = useWriteContract();

async function connectToWallet() {
  try {
    await connect({ connector: injected() });
    alert('MetaMask connected successfully!');
  } catch (error) {
    console.error('Failed to connect MetaMask:', error);
    alert('Error connecting to MetaMask');
  }
}

async function sendTransaction() {
  if (!connected.value) {
    alert('Please connect to MetaMask first.');
    return;
  }

  try {
    if (!/^0x[a-fA-F0-9]{40}$/.test(recipient.value)) {
      alert('Invalid recipient address');
      return;
    }

    if (Number(amount.value) <= 0) {
      alert('Amount must be greater than 0');
      return;
    }

    // Send the transaction
    const tx = await writeContractAsync({
      address: usdtContractAddress,
      abi: USDT_ABI,
      functionName: 'transfer',
      args: [recipient.value, BigInt(amount.value * 10 ** 6)], // USDT has 6 decimals
      chainId: 1, // Ethereum Mainnet
    });

    console.log('Transaction sent:', tx);
    alert('Transaction successfully sent. Waiting for confirmation...');
  } catch (error) {
    console.error('Transaction failed:', error);
    alert(`Transaction failed: ${error.message || 'Unknown error'}`);
  }
}`
